src/commands.jl

```diff
@@ -1,47 +1,57 @@
 function perform_return!(state::DebuggerState)
-    returning_frame = state.stack[end]
+    returning_frame = state.frame
     returning_expr = pc_expr(returning_frame)
     @assert isexpr(returning_expr, :return)
     val = @lookup(returning_frame, returning_expr.args[1])
-    if length(state.stack) != 1
-        calling_frame = state.stack[end-1]
+    if !isempty(state.stack)
         if returning_frame.code.generator
             # Don't do anything here, just return us to where we were
         else
-            prev = pc_expr(calling_frame)
+            state.frame = pop!(state.stack)
+            prev = pc_expr(state.frame)
             if isexpr(prev, :(=))
-                do_assignment!(calling_frame, prev.args[1], val)
-            elseif isassign(calling_frame)
-                do_assignment!(calling_frame, getlhs(calling_frame.pc[]), val)
+                do_assignment!(state.frame, prev.args[1], val)
+            elseif isassign(state.frame)
+                do_assignment!(state.frame, getlhs(state.frame.pc[]), val)
             end
-            state.stack[end-1] = JuliaStackFrame(calling_frame, maybe_next_call!(Compiled(), calling_frame,
-                calling_frame.pc[] + 1))
+            state.frame.pc[] += 1
+            maybe_next_call!(state.stack, state.frame)
         end
     else
         @assert !returning_frame.code.generator
         state.overall_result = val
     end
-    pop!(state.stack)
-    if !isempty(state.stack) && state.stack[end].code.wrapper
-        finish!(Compiled(), state.stack[end])
+    if !isempty(state.stack) && state.frame.code.wrapper
+        finish!(state.stack, state.frame)
         perform_return!(state)
     end
 end
 
 function propagate_exception!(state::DebuggerState, exc)
-    while !isempty(state.stack)
-        pop!(state.stack)
-        isempty(state.stack) && break
-        if isa(state.stack[end], JuliaStackFrame)
-            if !isempty(state.stack[end].exception_frames)
-                # Exception caught
-                state.stack[end] = JuliaStackFrame(state.stack[end],
-                    JuliaProgramCounter(state.stack[end].exception_frames[end]))
-                state.stack[end].last_exception[] = exc
-                return true
-            end
+    original_stack = copy(state.stack)
```

Comment:
For example here, just being able to store `original_frame = stack.frame` and then restore `stack.frame = original_frame` would be nice, instead of having to both store the frame and the stack to restore the original state.